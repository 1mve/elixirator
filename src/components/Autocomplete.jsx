import React, { useState, useMemo, useRef, useEffect } from 'react';
import { getSuggestions } from '../utils/utils';

const Autocomplete = ({ list, limit }) => {
    const [searchString, setSearchString] = useState('');
    const [isVisible, setIsVisible] = useState(false);
    const listWrap = useRef(null);

    useEffect(() => {
        document.addEventListener('click', handleClickOutside);

        return () => document.removeEventListener('click', handleClickOutside);
    }, []);

    const handleInputChange = ({ target }) => {
        setSearchString(target.value);
    };

    const handleClickOutside = event => {
        const { current } = listWrap;

        if (current && !current.contains(event.target)) {
            setIsVisible(false);
        }
    };

    const handleInputClick = () => {
        if (!searchString) {
            setIsVisible(!isVisible);
        }
    };

    const suggestionsList = useMemo(() =>
        getSuggestions({ list, limit, searchString })
    , [searchString, list, limit]);
    
    return (
        <div className="autocomplete-container" ref={listWrap}>
            <h2>Autocomplete by Misha Velmik</h2>
            <input
                placeholder="Type here to search (try with numbers)"
                value={searchString}
                onChange={(handleInputChange)}
                onClick={handleInputClick}
            />
            {isVisible && (
                <div className="list">
                    {suggestionsList
                        .map(({ id, value }) => (
                            <div
                                key={id}
                                onClick={() => setSearchString(value)}
                                className="list-item"
                            >
                                {value}
                            </div>
                        ))
                    }
                </div>
            )}
        </div>
    )
};

export default Autocomplete;