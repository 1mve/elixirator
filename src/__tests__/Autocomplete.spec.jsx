import React from 'react';
import { shallow } from 'enzyme';
import Autocomplete from '../components/Autocomplete';
import renderer from 'react-test-renderer';

describe('<Autocomplete />', () => {
    let props;

    beforeEach(() => {
        props = {
            limit: 25,
            list: [
                {
                    id: 'id-1',
                    value: 'value-1'
                },
                {
                    id: 'id-2',
                    value: 'value-2'
                },
                {
                    id: 'id-3',
                    value: 'value-3'
                },
                {
                    id: 'id-4',
                    value: 'value-4'
                },
                {
                    id: 'id-5',
                    value: 'value-5'
                },
                {
                    id: 'id-6',
                    value: 'value-6'
                },
                {
                    id: 'id-7',
                    value: 'value-7'
                },
                {
                    id: 'id-8',
                    value: 'value-8'
                },
                {
                    id: 'id-9',
                    value: 'value-9'
                },
                {
                    id: 'id-10',
                    value: 'value-10'
                }
            ]
        };
    });

    it('should NOT show suggestions if no search string typed', () => {
        const wrapper = shallow(
            <Autocomplete {...props} />,
        );

        expect(wrapper.find('.list').exists()).toBe(false);
    });

    it('should show 10 suggestions if no search string typed but click on empty input once', () => {
        const wrapper = shallow(
            <Autocomplete {...props} />,
        );

        wrapper.find('input').simulate('click');

        expect(wrapper.find('.list-item').length).toEqual(10);
    });

    it('should show 3 suggestions if no search string typed but click on empty input once and limit is 3', () => {
        const testLimit = 3;
        const wrapper = shallow(
            <Autocomplete {...props} limit={testLimit} />,
        );

        wrapper.find('input').simulate('click');

        expect(wrapper.find('.list-item').length).toEqual(testLimit);
    });

    it('should show 2 suggestions if input`s value is `value-1`', () => {
        const wrapper = shallow(
            <Autocomplete {...props} />,
        );

        wrapper.find('input').simulate('click');
        wrapper.find('input').simulate('change', { target: { value: 'value-1' } });

        expect(wrapper.find('.list-item').length).toEqual(2);
    });

    it('should show match snapshot if input`s value is `value-1`', () => {
        const wrapper = shallow(
            <Autocomplete {...props} />,
        );

        wrapper.find('input').simulate('click');
        wrapper.find('input').simulate('change', { target: { value: 'value-1' } });

        expect(wrapper).toMatchSnapshot();
    });
});