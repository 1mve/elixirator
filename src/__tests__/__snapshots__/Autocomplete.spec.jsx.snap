// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Autocomplete /> should show match snapshot if input\`s value is \`value-1\` 1`] = `
<div
  className="autocomplete-container"
>
  <h2>
    Autocomplete by Misha Velmik
  </h2>
  <input
    onChange={[Function]}
    onClick={[Function]}
    placeholder="Type here to search (try with numbers)"
    value="value-1"
  />
  <div
    className="list"
  >
    <div
      className="list-item"
      key="id-1"
      onClick={[Function]}
    >
      value-1
    </div>
    <div
      className="list-item"
      key="id-10"
      onClick={[Function]}
    >
      value-10
    </div>
  </div>
</div>
`;
